lib/WasmReader/WasmBinaryReader.cpp

```diff
@@ -60,157 +69,208 @@ WasmBinaryReader::WasmBinaryReader(PageAllocator * alloc, byte* source, size_t l
     ResetModuleData();
 }
 
+void WasmBinaryReader::InitializeReader()
+{
+    ModuleHeader();
+#if DBG
+    if (m_trace)
+    {
+        byte* startModule = m_pc;
+
+        bool doRead = true;
+        SectionCode prevSect = bSectInvalid;
+        while (doRead)
+        {
+            SectionHeader secHeader = ReadSectionHeader();
+            if (secHeader.code <= prevSect)
+            {
+                TRACE_WASM_DECODER(L"Unknown section order");
+            }
+            prevSect = secHeader.code;
+            // skip the section
+            m_pc = secHeader.end;
+            doRead = !EndOfModule() && secHeader.code != bSectEnd;
+        }
+        m_pc = startModule;
+    }
+#endif
+}
+
+void
+WasmBinaryReader::ThrowDecodingError(const wchar_t* msg, ...)
+{
+    va_list argptr;
+    va_start(argptr, msg);
+    Output::Print(L"Binary decoding failed: ");
+    Output::VPrint(msg, argptr);
+    Output::Flush();
+    Js::Throw::InternalError();
+}
+
 bool
 WasmBinaryReader::IsBinaryReader()
 {
     return true;
 }
 
-WasmOp
-WasmBinaryReader::ReadFromModule()
+bool 
+WasmBinaryReader::ReadNextSection(SectionCode nextSection)
 {
-    TRACE_WASM_DECODER(L"Decoding Module");
+    if (EndOfModule() || sectionFlags[nextSection] == fSectIgnore)
+    {
+        return false;
+    }
 
-    SectionCode sectionId;
-    UINT length = 0;
+    // TODO: Should check that previous section, if any, was of the correct
+    // length by comparing against m_moduleState.byteLen.
```

Comment:
Stale comment referencing m_moduleState.byteLen which is now redundant with secHeader.start and secHeader.end.